### ISSUE TITLE
Nd softmax

### ISSUE THREAD

**Issue by dfdx** [FluxML/NNlib.jl#18]
In NNlib, we have 1D/2D softmax, in CUDNN we have 4D softmax, so essentially we don't have any that works for both - CPU and GPU. 

More generally, we may want to provide softmax implementation with configurable dimension to scale (e.g. like [in PyTorch](http://pytorch.org/docs/master/nn.html#torch.nn.Softmax)).

**Comment by jekbradbury** [comment]
Something like
```julia
function softmax(x, axis=1)
    out = exp.(x .- maximum(x, axis))
    out ./= sum(out, axis)
    return out
end
```
for the generic implementation, plus a CUDA kernel based on [PyTorch's](https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/SoftMaxCommon.cuh#L389) (because theirs is faster than cuDNN and has configurable dimension)?

**Comment by MikeInnes** [comment]
See https://github.com/JuliaGPU/CuArrays.jl/issues/45. We can update the CPU api once we have that kernel.
